### ISSUE TITLE
support multiple ldap servers

### ISSUE THREAD

**Issue by 4nd3r** [matrix-org/matrix-synapse-ldap3#11]
uri `ldaps://ldap1.example.com ldap2.example.com/` etc?

**Comment by mweinelt** [comment]
I'd rather extend uri to support being a list.

**Comment by mweinelt** [comment]
@4nd3r This might already be working if you pass a list. Can you give it a try?

**Comment by 4nd3r** [comment]
nope

```
2016-12-20 17:55:51,492 - synapse.http.server - 139 - ERROR - POST-17- Failed handle request synapse.http.server._async_render on <synapse.rest.ClientRestResource object at 0x7f475634d790>: <SynapseRequest at 0x7f4755e17f80 method=POST uri=/_matrix/client/r0/login? clientproto=HTTP/1.1 site=8448>                 
Traceback (most recent call last):                                                                                                                             
File "/usr/lib/python2.7/dist-packages/synapse/http/server.py", line 116, in wrapped_request_handler                                                       
    yield request_handler(self, request, request_metrics)                                                                                                    
AttributeError: 'list' object has no attribute 'startswith'                                                                               
```

**Comment by mweinelt** [comment]
Huh, can you shed some light on this @erikjohnston?

**Comment by mweinelt** [comment]
Are you already using this module? Does your config look like this? Did you change the uri below this path? Which synapse version are you using?

```
password_providers:
 - module: "ldap_auth_provider.LdapAuthProvider"
   config:
    enabled: true
    uri: 
      - "ldaps://ldap1.example.com"
      - "ldaps://ldap2.example.com
[...]
```

**Comment by jcgruenhage** [comment]
Shouldn't that work with just adding the ldap provider module twice to the password_providers list twice with the different servers?

**Comment by mweinelt** [comment]
I think it sounds like it's worth a try.

**Comment by jcgruenhage** [comment]
I tried exactly that and it didn't work for me, it did not work with a single one either, so there probably is just another problem in my setup :D

**Comment by sim0nx** [comment]
> Are you already using this module? Does your config look like this? Did you change the uri below this path? Which synapse version are you using?
> 
> ```
> password_providers:
>  - module: "ldap_auth_provider.LdapAuthProvider"
>    config:
>     enabled: true
>     uri: 
>       - "ldaps://ldap1.example.com"
>       - "ldaps://ldap2.example.com"
> [...]
> ```

HI,

I am using latest synapse (1.10) with the latest ldap module version and your suggested format does not work.
In the logs I see something like "strip()" does not work on "type list".
Using something like "uri: 'server1, server2, ..." does not work either.

**Comment by jcgruenhage** [comment]
Have you tried my suggestion of duplicating the whole block? 
```
password_providers:
 - module: "ldap_auth_provider.LdapAuthProvider"
   config:
    enabled: true
    uri: 
      - "ldaps://ldap1.example.com"
    [...]
 - module: "ldap_auth_provider.LdapAuthProvider"
   config:
    enabled: true
    uri: 
      - "ldaps://ldap2.example.com"
    [...]
```

**Comment by sim0nx** [comment]
> Have you tried my suggestion of duplicating the whole block?
> 
> ```
> password_providers:
>  - module: "ldap_auth_provider.LdapAuthProvider"
>    config:
>     enabled: true
>     uri: 
>       - "ldaps://ldap1.example.com"
>     [...]
>  - module: "ldap_auth_provider.LdapAuthProvider"
>    config:
>     enabled: true
>     uri: 
>       - "ldaps://ldap2.example.com"
>     [...]
> ```

No sorry  I did not read that part... that seems to work fine! Thank you

**Comment by jcgruenhage** [comment]
I think we can close this issue then.

**Comment by army1349** [comment]
> 
> 
> I think we can close this issue then.

Please don't do that and consider [PR](https://github.com/matrix-org/matrix-synapse-ldap3/pull/85).

**Comment by jcgruenhage** [comment]
A right, that's not the usecase I had in mind. I had multiple authentication sources in mind, not one highly available pool. In that case, yes, the PR looks good.

**Comment by richvdh** [comment]
fixed by #85
